igvm/hypervisor_preferences.py

```diff
@@ -1,63 +1,93 @@
+return either a score of range 0.0 - 1.0 inclusive, or a boolean value, which
+is only there for convenience and converts to float.
+A higher score means that a Hypervisor should be favored, while a low one
+discourages a HV. If False or 0.0 is given, that means a HV does not fulfill
+the requirements at all and is excluded.
+
+Copyright (c) 2020 InnoGames GmbH
 """
-# This module contains the preferences as simple classes.  We try to keep
-# them reusable, even though most of them are not reused.  Some of the classes
-# are so simple that they could as well just be a function, but kept
-# as classes to have a consistent style.
-
 from logging import getLogger
-
-from igvm.utils import LazyCompare
+from typing import Union, List
 
 log = getLogger(__name__)
 
 
-class InsufficientResource(object):
-    """Check a resource of hypervisor would be sufficient"""
-    def __init__(self, hv_attribute, vm_attribute, multiplier=1, reserved=0):
-        self.hv_attribute = hv_attribute
-        self.vm_attribute = vm_attribute
-        self.multiplier = multiplier  # TODO Use identical units in Serveradmin
-        self.reserved = reserved
-
-    def __repr__(self):
+class HypervisorPreference:
+    """The base class for all HV preferences."""
+
+    def get_score(self, vm, hv) -> Union[float, bool]:
+        """Calculates a preference value to indicate how good a HV fits.
+
+        @param vm: The VM object to check against the HV.
+        @param hv: The HV object to check against the VM.
+        @return: A float or a bool. The bool is just for convenience and will
+                 be converted to a float. A value of 0.0 means the HV does not
+                 fulfill the requirement at all. But a value of 1.0 means it is
+                 a very good candidate for this particular preference.
+        @rtype: Union[float, bool]
+        """
+        raise NotImplementedError('get_preference is not implemented')
+
+
+class InsufficientResource(HypervisorPreference):
+    """Check whether a resource of a hypervisor would be sufficient."""
+
+    def __init__(
+        self,
+        hv_attribute: str,
+        vm_attribute: str,
+        multiplier: int = 1,
+        reserved: int = 0,
+    ) -> None:
+        self.hv_attribute: str = hv_attribute
+        self.vm_attribute: str = vm_attribute
+        # TODO Use identical units in Serveradmin
+        self.multiplier: int = multiplier
+        self.reserved: int = reserved
+
+    def __repr__(self) -> str:
         args = repr(self.hv_attribute)
         if self.reserved:
             args += ', reserved=' + repr(self.reserved)
 
         return '{}({})'.format(type(self).__name__, args)
 
-    def __call__(self, vm, hv):
+    def get_score(self, vm, hv) -> Union[float, bool]:
         # Treat freshly created HVs always passing this check
         if not hv.dataset_obj[self.hv_attribute]:
-            return False
+            return True
```

Comment:
```suggestion
            # Because new installed Hypervisors don't have values 
            # (e.g. for load) - they are fresh meat ...
            return True
```